### ISSUE TITLE
Performance slows down by using multi-root workspace

### ISSUE THREAD

**Issue by olafkrueger** [BowlerHatLLC/vscode-nextgenas#255]
I switched back to using "one VSCode instance for a particular project/lib" because multi-root is not usable for me because of performance issues.
E.g. I can watch how the the red swiggles grow character by character during typing or even after finishing typing. Code completition needs a couple of seconds.
(Restarting the language server or reloading the window doesn't help)

Morever I notice weird/wrong error messages and warnings which is not the case when using a single VSCode instance with a single project. (Maybe the reason for this is the bad performance...?)

I am using the latest VSCode and Extension versions on Windows7.

Before diving deeper into this (I'll checkout how it works on another machine) I'd would like to ask if others also ran into this issue... maybe my machine is the issue ;-)

Thanks,
Olaf

**Comment by joshtynjala** [comment]
Can you open your multi-root worried and work inside it until it starts performing poorly? Then, open VSCode's "Output" view and switch to "ActionScript & MXML: Language Server". If you see some error logs for the current session, copy/paste them here. That might give is a clue.

**Comment by joshtynjala** [comment]
How many root folders do you have in the workspace?

**Comment by olafkrueger** [comment]
Hi Josh,
forgot to mention that the "Output" is full of errors.
This is an extract of it [1].
Does this help?

Just to avoid confusion:
I noticed that the "royale compiler" is mentioned somewhere in the logs: I am just using the Flex SDK.

Thanks!
Olaf


[1] Output
```
Aug 22, 2018 11:09:48 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNUNG: Unmatched cancel notification for request id 1
```

```
Aug 22, 2018 11:24:56 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
WARNUNG: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "textDocument/didChange",
  "params": {
    "textDocument": {
      "version": 108,
      "uri": "file:///d%3A/Daten/github/ebase4x/client/src/...
    },
    "contentChanges": [
      {
        "range": {
          "start": {
            "line": 18,
            "character": 17
          },
          "end": {
            "line": 18,
            "character": 17
          }
        },
        "rangeLength": 0,
        "text": "t"
      }
    ]
  }
}
```

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:63)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:148)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:216)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:183)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:188)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:90)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:95)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor7.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:61)
	... 11 more
Caused by: java.lang.NullPointerException
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.getCompilationUnitForDefinition(ASProjectScope.java:1364)
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.removeDefinition(ASProjectScope.java:1127)
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.removeCompilationUnits(ASProjectScope.java:1410)
	at org.apache.royale.compiler.internal.units.CompilationUnitBase.clean(CompilationUnitBase.java:721)
	at org.apache.royale.compiler.internal.workspaces.Workspace.invalidate(Workspace.java:410)
	at org.apache.royale.compiler.internal.workspaces.Workspace.fileChanged(Workspace.java:525)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.didChange(ActionScriptTextDocumentService.java:1578)
	... 15 more
```

```Aug 22, 2018 2:03:45 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNUNG: Unmatched cancel notification for request id 928
Aug 22, 2018 2:08:41 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SCHWERWIEGEND: Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:63)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:116)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:257)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:186)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:188)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:90)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:95)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor12.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:61)
	... 11 more
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2214)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4053)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4057)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4986)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4992)
	at org.apache.royale.compiler.internal.projects.CompilerProject.getCacheForScope(CompilerProject.java:676)
	at org.apache.royale.compiler.internal.scopes.ASScope.findPropertyQualified(ASScope.java:1592)
	at org.apache.royale.compiler.internal.definitions.references.ResolvedQualifiersReference.resolve(ResolvedQualifiersReference.java:80)
	at org.apache.royale.compiler.internal.definitions.DefinitionBase.resolveType(DefinitionBase.java:1055)
	at org.apache.royale.compiler.internal.definitions.DefinitionBase.resolveType(DefinitionBase.java:1095)
	at org.apache.royale.compiler.internal.definitions.ClassDefinition.resolveBaseClass(ClassDefinition.java:362)
	at org.apache.royale.compiler.internal.definitions.TypeDefinitionBase$TypeIterator.pushChildren(TypeDefinitionBase.java:283)
	at org.apache.royale.compiler.internal.definitions.TypeDefinitionBase$TypeIterator.next(TypeDefinitionBase.java:217)
	at org.apache.royale.compiler.internal.definitions.TypeDefinitionBase$TypeIterator.next(TypeDefinitionBase.java:175)
	at org.apache.royale.compiler.internal.scopes.TypeScope.getPropertyForScopeChain(TypeScope.java:262)
	at org.apache.royale.compiler.internal.scopes.ScopeView.getPropertyForScopeChain(ScopeView.java:72)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:1119)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:1061)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:1277)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:999)
	at org.apache.royale.compiler.internal.scopes.ASScopeCache.findProperty(ASScopeCache.java:153)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:1428)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolve(IdentifierNode.java:394)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveType(IdentifierNode.java:466)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveMemberRef(IdentifierNode.java:885)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolve(IdentifierNode.java:388)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveType(IdentifierNode.java:466)
	at org.apache.royale.compiler.internal.tree.as.MemberAccessExpressionNode.resolveType(MemberAccessExpressionNode.java:136)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveMemberRef(IdentifierNode.java:885)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolve(IdentifierNode.java:388)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveType(IdentifierNode.java:466)
	at org.apache.royale.compiler.internal.tree.as.MemberAccessExpressionNode.resolveType(MemberAccessExpressionNode.java:136)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveMemberRef(IdentifierNode.java:885)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolve(IdentifierNode.java:388)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveType(IdentifierNode.java:466)
	at org.apache.royale.compiler.internal.tree.as.MemberAccessExpressionNode.resolveType(MemberAccessExpressionNode.java:136)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolveMemberRef(IdentifierNode.java:885)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolve(IdentifierNode.java:388)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.actionScriptHoverWithNode(ActionScriptTextDocumentService.java:2794)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.actionScriptHover(ActionScriptTextDocumentService.java:2775)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.hover(ActionScriptTextDocumentService.java:471)
	... 15 more
Caused by: java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:770)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4052)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4057)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4986)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4992)
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.addScopeToCompilationUnitScopeList(ASProjectScope.java:1310)
	at org.apache.royale.compiler.internal.projects.CompilerProject$ScopeCacheLoader.load(CompilerProject.java:107)
	at org.apache.royale.compiler.internal.projects.CompilerProject$ScopeCacheLoader.load(CompilerProject.java:96)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3628)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2336)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2295)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2208)
	... 55 more
```

```
[Error - 4:12:03 PM] Request textDocument/hover failed.
  Message: Internal error.
  Code: -32603 
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:63)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:116)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:257)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:186)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:188)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:90)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:95)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor12.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:61)
	... 11 more
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2214)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4053)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4057)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4986)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4992)
	at org.apache.royale.compiler.internal.projects.CompilerProject.getCacheForScope(CompilerProject.java:676)
	at org.apache.royale.compiler.internal.scopes.ASScope.findPropertyQualified(ASScope.java:1592)
	at org.apache.royale.compiler.internal.definitions.references.ResolvedQualifiersReference.resolve(ResolvedQualifiersReference.java:80)
	at org.apache.royale.compiler.internal.definitions.DefinitionBase.resolveType(DefinitionBase.java:1055)
	at org.apache.royale.compiler.internal.definitions.DefinitionBase.resolveType(DefinitionBase.java:1095)
	at org.apache.royale.compiler.internal.definitions.ClassDefinition.resolveBaseClass(ClassDefinition.java:362)
	at org.apache.royale.compiler.internal.definitions.ClassDefinitionBase$ClassIterator.next(ClassDefinitionBase.java:512)
	at org.apache.royale.compiler.internal.definitions.ClassDefinitionBase$ClassIterator.next(ClassDefinitionBase.java:478)
	at org.apache.royale.compiler.internal.scopes.TypeScope.addImplicitOpenNamespaces(TypeScope.java:557)
	at org.apache.royale.compiler.internal.scopes.ScopeView.addImplicitOpenNamespaces(ScopeView.java:115)
	at org.apache.royale.compiler.internal.scopes.ASScope.getNamespaceSetImpl(ASScope.java:558)
	at org.apache.royale.compiler.internal.scopes.ASScopeCache.getNamespaceSet(ASScopeCache.java:395)
	at org.apache.royale.compiler.internal.scopes.ASScope.getNamespaceSet(ASScope.java:515)
	at org.apache.royale.compiler.internal.scopes.ASScope.addNamespacesFromContainingScope(ASScope.java:578)
	at org.apache.royale.compiler.internal.scopes.ASScope.getNamespaceSetImpl(ASScope.java:561)
	at org.apache.royale.compiler.internal.scopes.ASScopeCache.getNamespaceSet(ASScopeCache.java:395)
	at org.apache.royale.compiler.internal.scopes.ASScope.getNamespaceSet(ASScope.java:515)
	at org.apache.royale.compiler.internal.scopes.ASScope.getNamespaceSetForNameImpl(ASScope.java:484)
	at org.apache.royale.compiler.internal.scopes.ASScopeCache.getNamespaceSetForName(ASScopeCache.java:557)
	at org.apache.royale.compiler.internal.scopes.ASScope.getNamespaceSetForName(ASScope.java:447)
	at org.apache.royale.compiler.internal.scopes.ASScopeCache.findProperty(ASScopeCache.java:151)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:1428)
	at org.apache.royale.compiler.internal.scopes.ASScope.findProperty(ASScope.java:1423)
	at org.apache.royale.compiler.internal.tree.as.IdentifierNode.resolve(IdentifierNode.java:394)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.actionScriptHoverWithNode(ActionScriptTextDocumentService.java:2794)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.actionScriptHover(ActionScriptTextDocumentService.java:2775)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.hover(ActionScriptTextDocumentService.java:471)
	... 15 more
Caused by: java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:770)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4052)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4057)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4986)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4992)
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.addScopeToCompilationUnitScopeList(ASProjectScope.java:1310)
	at org.apache.royale.compiler.internal.projects.CompilerProject$ScopeCacheLoader.load(CompilerProject.java:107)
	at org.apache.royale.compiler.internal.projects.CompilerProject$ScopeCacheLoader.load(CompilerProject.java:96)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3628)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2336)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2295)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2208)
	... 46 more
```

**Comment by olafkrueger** [comment]
> How many root folders do you have in the workspace?

12 root folders

**Comment by joshtynjala** [comment]
> This is an extract of it [1].
> Does this help?

Thanks! I've seen some of those errors before, but I'm not quite sure how to fix them. This error is new:

```
Caused by: java.lang.NullPointerException
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.getCompilationUnitForDefinition(ASProjectScope.java:1364)
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.removeDefinition(ASProjectScope.java:1127)
	at org.apache.royale.compiler.internal.scopes.ASProjectScope.removeCompilationUnits(ASProjectScope.java:1410)
	at org.apache.royale.compiler.internal.units.CompilationUnitBase.clean(CompilationUnitBase.java:721)
	at org.apache.royale.compiler.internal.workspaces.Workspace.invalidate(Workspace.java:410)
	at org.apache.royale.compiler.internal.workspaces.Workspace.fileChanged(Workspace.java:525)
	at com.as3mxml.vscode.ActionScriptTextDocumentService.didChange(ActionScriptTextDocumentService.java:1578)
```

I just looked at the source code for the Royale compiler, and I think there may be a bug that I can fix.

>  I noticed that the "royale compiler" is mentioned somewhere in the logs: I am just using the Flex SDK.

The Royale compiler always powers the code intelligence in the VSCode extension. It is compatible with all SDKs.

> 12 root folders

Wow! That's way more than I expected. To be honest, I'm not surprised that there are some performance issues with 12 root folders.

**Comment by joshtynjala** [comment]
Fixed the exception that I mentioned in my previous comment.

apache/royale-compiler@f388f882c9e5f6ee282bbf545ae546ea28ac28b6

I'm hoping to see another Royale release soon because I also made several other compiler fixes over the last few months that will benefit the VSCode extension.

**Comment by piotrzarzycki21** [comment]
I believe reducing root folders is not a solution. 12 modules opened at a time is not so much. The question is what more information should we provide to give you ability for improvements ? Does fix in Royale compiler help with above issue ?

**Comment by joshtynjala** [comment]
> The question is what more information should we provide to give you ability for improvements ?

The thing that will help me most is determining reliable ways to reproduce errors that show up in the Output panel.

**Comment by joshtynjala** [comment]
Fixing issue #253 may improve performance too. Right now, the extension does not properly handle requests from VSCode that are cancelled. It doesn't hurt to let them fully complete, but it should be faster if I can abort early when a cancelled request is detected.

**Comment by piotrzarzycki21** [comment]
@olafkrueger maybe let's not close that issue so soon yet. Let's discuss it internally. I believe it's time to make some effort and release next version of Royale.

**Comment by joshtynjala** [comment]
I've figured out that exceptions thrown from `com.google.common` classes are caused by threads getting out of sync. Sometimes, everything needs to pause until certain tasks complete, and that's not happening in all cases.

In the process of fixing #253, I think I figured out how to avoid these exceptions. I don't know if this will help performance for workspaces with many root folders, but it will be a stability improvement at least.

**Comment by joshtynjala** [comment]
Alright, over the last couple of days, I've done some heavy refactoring that I think will fix the performance issues with a large number of workspace folders. To test, I just tried a workspace containing 16 examples from Feathers, and it seems to be performing well. In fact, it should improve performance in single-folder workspaces too! I finally figured out how to handle multi-threading with the Royale compiler, and everything seems more responsive now.

I'm going to close this issue as resolved. After I release v0.14, if you find that you are still having performance issues, we can reopen and investigate some more.
